### PR TITLE
Add workaround to handle zero sized files in /proc

### LINKED ISSUE #3410: `File#read` does not read files located in `/proc` Linux Virtual FileSystem
From **TLDP** [Linux Filesystem Hierarchy: /proc](http://www.tldp.org/LDP/Linux-Filesystem-Hierarchy/html/proc.html)
> The most distinctive thing about files in this directory is the fact that all of them have a file size of 0, with the exception of kcore, mtrr and self.

With **Ruby**
```Ruby
puts File.read("/proc/self/cmdline").empty? # => false
````

With **Crystal**
```Crystal
if File.directory?("/proc/self")
  path = "/proc/self/cmdline"

  puts <<-EOF
  Given #{path.inspect} file path
  File#read empty?        # => #{File.read(path).size == 0}
  File.gets_to_end empty? # => #{File.open(path, &.gets_to_end).size == 0}
  EOF
end
```

Output:
```:
Given "/proc/self/cmdline" file path
File#read empty?        # => true
File.gets_to_end empty? # => false
```

### PR DESCRIPTION
**Pull request by Val**
Fix #3410

### CODE REVIEW

File: src/file.cr [hunk 1]
```diff
@@ -417,6 +417,11 @@ class File < IO::FileDescriptor
   # ```
   def self.read(filename, encoding = nil, invalid = nil) : String
     File.open(filename, "r") do |file|
+      {% if flag?(:linux) %}
+      # workaround to handle zero sized files in /proc Linux virtual filesystem
+      encoding ||= "UTF-8" if File.expand_path(filename) =~ /^\/proc/
```
Comment (RX14) [inline src/file.cr]: Why not use `starts_with?`? Regexes are quite slow and unnecessary for this.

**Comment by asterite** [comment]
I would fix this by changing `File.read` to open the file and invoke `gets_to_end` when the file size is zero, or always. The problem has nothing to do with encoding.

File: spec/std/file_spec.cr [hunk 2]
```diff
@@ -37,6 +37,13 @@ describe "File" do
     str.should eq("Hello World\n" * 20)
   end
 
+  {% if !flag?(:windows) %}
```
Comment (RX14) [inline spec/std/file_spec.cr]: Yeah, `/proc` is only pretty much guaranteed to be mounted on windows, but it *can* be mounted on most unixes.

File: src/file.cr [hunk 3]
```diff
@@ -417,6 +417,10 @@ class File < IO::FileDescriptor
   # ```
   def self.read(filename, encoding = nil, invalid = nil) : String
     File.open(filename, "r") do |file|
+      {% if flag?(:linux) %}
+      return file.gets_to_end if File.expand_path(filename).starts_with?("/proc")
```
Comment (RX14) [inline src/file.cr]: This now ignores encoding for `/proc*`.

**Comment by asterite** [comment]
@Val we finally decided with @waj that the safest thing to do, even if it's a bit slower, is to always invoke `gets_to_end`. So basically it should be:

```cr
  def self.read(filename, encoding = nil, invalid = nil) : String
    File.open(filename, "r") do |file|
      file.set_encoding(encoding, invalid: invalid) if encoding
      file.gets_to_end
    end
  end
```

But I'll let you change this in your PR because you have the specs, and you found the bug :-)

**Comment by RX14** [comment]
@asterite I think `File.read` is a common enough call that optimising it is worthwhile.
I think we should at least do some benchmarks to determine the slowdown factor before we merge this.

**Comment by Val** [comment]
```Crystal
path = "/proc/self/cmdline"

def read_short(filename, encoding = nil, invalid = nil) : String
  File.open(filename, "r") do |file|
    if encoding
      file.set_encoding(encoding, invalid: invalid)
      file.gets_to_end
    else
      {% if flag?(:linux) %}
      return file.gets_to_end if File.expand_path(filename).starts_with?("/proc")
      {% end %}

      size = file.size.to_i
      String.new(size) do |buffer|
        file.read Slice.new(buffer, size)
        {size.to_i, 0}
      end
    end
  end
end

def read_safe(filename, encoding = nil, invalid = nil) : String
  File.open(filename, "r") do |file|
    file.set_encoding(encoding, invalid: invalid) if encoding
    file.gets_to_end
  end
end

require "benchmark"
Benchmark.ips do |x|
  x.report("read short") { read_short path }
  x.report("read safe")  { read_safe  path }
end
```

```
>  crystal build file_read_bench.rb
> ./file_read_bench
read1 174.99k (± 6.28%)  1.73× slower
read2 301.93k (± 5.34%)       fastest
> ./file_read_bench
read1 212.89k (± 5.30%)  1.24× slower
read2 263.87k (± 3.41%)       fastest
```

**Comment by RX14** [comment]
@Val can you also test on various sizes of normal file, say 1kb, 1mb,10mb?

**Comment by Val** [comment]
```Crystal
path = "/tmp/#{File.basename(__FILE__)}"

def read1(filename, encoding = nil, invalid = nil) : String
  File.open(filename, "r") do |file|
    if encoding
      file.set_encoding(encoding, invalid: invalid)
      file.gets_to_end
    else
      {% if flag?(:linux) %}
      return file.gets_to_end if File.expand_path(filename).starts_with?("/proc")
      {% end %}

      size = file.size.to_i
      String.new(size) do |buffer|
        file.read Slice.new(buffer, size)
        {size.to_i, 0}
      end
    end
  end
end

def read2(filename, encoding = nil, invalid = nil) : String
  File.open(filename, "r") do |file|
    file.set_encoding(encoding, invalid: invalid) if encoding
    file.gets_to_end
  end
end

BLOBS = {} of Symbol => String
kb_01 = BLOBS[:kb_01] = "X" * 1024
mb_01 = BLOBS[:mb_01] = kb_01 * 1024
BLOBS[:mb_10] = mb_01 * 10

require "benchmark"

BLOBS.each do |name, blob|
  File.open(path, "w") { |f| f.print blob }

  puts "#{name} _____________________________________\n"
  Benchmark.ips do |x|
    x.report("read1 #{name}") { read1 path }
    x.report("read2 #{name}") { read2 path }
  end
  puts
end

File.delete path
```

Output
```
> crystal build --release file_read_bench.cr 
> ./file_read_bench 
kb_01 _____________________________________

read1 kb_01 349.94k (± 2.17%)       fastest
read2 kb_01 336.11k (± 2.58%)  1.04× slower

mb_01 _____________________________________

read1 mb_01   9.43k (± 2.27%)       fastest
read2 mb_01   2.32k (± 2.69%)  4.07× slower

mb_10 _____________________________________

read1 mb_10  532.3  (± 3.12%)       fastest
read2 mb_10 157.96  (± 1.78%)  3.37× slower
```

**Comment by Val** [comment]
Subtle, changing `return file.gets_to_end if File.expand_path(filename).starts_with?("/tmp")`.

```
> ./file_read_bench 
kb_01 _____________________________________

read1 kb_01 298.51k (± 2.54%)  1.13× slower
read2 kb_01 337.64k (± 3.03%)       fastest

mb_01 _____________________________________

read1 mb_01   2.32k (± 4.93%)       fastest
read2 mb_01   2.31k (± 6.39%)  1.01× slower

mb_10 _____________________________________

read1 mb_10 154.89  (± 5.72%)  1.04× slower
read2 mb_10 161.59  (± 4.46%)       fastest
```

**Comment by Val** [comment]
@RX14 & @asterite it seems that 416b0e7 should be reverted, what do you think?

**Pull request closed without merging**